cpp/src/io/comp/uncomp.cpp

```diff
@@ -616,4 +624,160 @@ std::vector<uint8_t> decompress(compression_type compression, host_span<uint8_t
   CUDF_FAIL("Unsupported compressed stream type");
 }
 
+void device_decompress(compression_type compression,
+                       device_span<device_span<uint8_t const> const> inputs,
+                       device_span<device_span<uint8_t> const> outputs,
+                       device_span<compression_result> results,
+                       size_t max_uncomp_chunk_size,
+                       size_t max_total_uncomp_size,
+                       rmm::cuda_stream_view stream)
+{
+  if (compression == compression_type::NONE) { return; }
+
+  auto const nvcomp_type      = to_nvcomp_compression(compression);
+  auto nvcomp_disabled_reason = nvcomp_type.has_value()
+                                  ? nvcomp::is_decompression_disabled(*nvcomp_type)
+                                  : "invalid compression type";
+  if (not nvcomp_disabled_reason) {
+    return nvcomp::batched_decompress(
+      *nvcomp_type, inputs, outputs, results, max_uncomp_chunk_size, max_total_uncomp_size, stream);
+  }
+
+  switch (compression) {
+    case compression_type::BROTLI: return gpu_debrotli(inputs, outputs, results, stream);
+    case compression_type::GZIP:
+      return gpuinflate(inputs, outputs, results, gzip_header_included::YES, stream);
+    case compression_type::SNAPPY: return gpu_unsnap(inputs, outputs, results, stream);
+    case compression_type::ZLIB:
+      return gpuinflate(inputs, outputs, results, gzip_header_included::NO, stream);
+    default: CUDF_FAIL("Compression error: " + nvcomp_disabled_reason.value());
+  }
+}
+
+void host_decompress(compression_type compression,
+                     device_span<device_span<uint8_t const> const> inputs,
+                     device_span<device_span<uint8_t> const> outputs,
+                     device_span<compression_result> results,
+                     rmm::cuda_stream_view stream)
+{
+  if (compression == compression_type::NONE) { return; }
+
+  auto const num_chunks = inputs.size();
+  auto const h_inputs   = cudf::detail::make_host_vector_async(inputs, stream);
+  auto const h_outputs  = cudf::detail::make_host_vector_async(outputs, stream);
+  stream.synchronize();
+
+  // Generate order vector to submit largest tasks first
+  std::vector<size_t> task_order(num_chunks);
+  std::iota(task_order.begin(), task_order.end(), 0);
+  std::sort(task_order.begin(), task_order.end(), [&](size_t a, size_t b) {
+    return h_inputs[a].size() > h_inputs[b].size();
+  });
+
+  std::vector<std::future<size_t>> tasks;
+  auto const num_streams =
+    std::min<std::size_t>({num_chunks,
+                           cudf::detail::global_cuda_stream_pool().get_stream_pool_size(),
+                           cudf::detail::host_worker_pool().get_thread_count()});
+  auto const streams = cudf::detail::fork_streams(stream, num_streams);
+  for (size_t i = 0; i < num_chunks; ++i) {
```

Comment:
```suggestion
  // Device input and output buffers are transferred to pinned memory allocations on the host. Note that if their total
  // size exceeds the pinned pool size, there may be a performance impact due to expensive new pinned allocations.
  auto const total_pinned_allocs = std::accumulate(inputs.begin(), inputs.end(), 0, [](size_type sum, auto dsp){
    return sum + dsp.size();
  }) + std::accumulate(inputs.begin(), inputs.end(), 0, [](size_type sum, auto dsp){
    return sum + dsp.size();
  });
  if(total_pinned_allocs > 100000000) 
    CUDF_LOG_WARN("new pinned allocation since total allocation (%zu) > pool size (%zu)", total_pinned_allocs, 100000000);
  for (size_t i = 0; i < num_chunks; ++i) {
```
Do you think it would useful to have a warning when we have to reallocate the pinned pool? I just set the pool size to 100MB, but if we use the env var to set it, it wouldn't be correct. How can we get the pool size?

Reply:
this computation isn't quite right, as we only keep the pinned buffers alive until the result has been copied to device. We have num_threads input and outputs allocated at any time, not all inputs/outputs.

Reply:
Oops, my bad. Thanks for clarifying!